File: ml-benchmark/autolifter/single-pass/cnt1s.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

cnt1s = fix (\f: Int -> Int -> List -> Int. \pre: Int. \cnt: Int. \xs: List.
  match xs with
    nil _ -> cnt
  | cons {h, t} ->
    let cnt = if and (== pre 0) (== h 1) then + cnt 1 else cnt in
      f h pre t
  end
) 0 0;

main = single_pass cnt1s;

File: ml-benchmark/autolifter/single-pass/cnt1s.ml
type lst = Nil of unit | Cons of int * lst

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let cnt1s =
  let rec f pre cnt xs =
    match xs with
    | Nil _ -> cnt
    | Cons (h, t) ->
      let cnt = if (pre = 0) && (h = 1) then cnt + 1 else cnt in
      f h pre t
  in
  f 0 0

let program = single_pass cnt1s